adam/learner/propose_but_verify.py

```diff
@@ -0,0 +1,372 @@
+from adam.ontology.phase1_spatial_relations import Region
+from adam.perception import ObjectPerception, PerceptionT, PerceptualRepresentation
+from adam.perception.developmental_primitive_perception import (
+    DevelopmentalPrimitivePerceptionFrame,
+    RgbColorPerception,
+)
+from adam.perception.perception_graph import (
+    DebugCallableType,
+    GraphLogger,
+    PerceptionGraph,
+    PerceptionGraphPattern,
+)
+from adam.utils import networkx_utils
+from attr import Factory, attrib, attrs
+
+
+@attrs()
+class ProposeButVerifyLanguageLearner(
+    Generic[PerceptionT, LinguisticDescriptionT],
+    LanguageLearner[PerceptionT, LinguisticDescription],
+):
+
+    _ontology: Ontology = attrib(validator=instance_of(Ontology), kw_only=True)
+    _observation_num = attrib(init=False, default=0)
+    _rng: Random = attrib(validator=instance_of(Random))
+    _words_to_hypotheses: Dict[str, PerceptionGraphPattern] = attrib(
+        init=False, default=Factory(dict)
+    )
+    _log_word_hypotheses_to: Optional[Path] = attrib(
+        validator=optional(instance_of(Path)), default=None
+    )
+    _graph_logger: Optional[GraphLogger] = attrib(
+        validator=optional(instance_of(GraphLogger)), default=None
+    )
+    debug_counter = 0
+    _debug_callback: Optional[DebugCallableType] = attrib(default=None)
+    _graph_match_confirmation_threshold: float = attrib(default=0.9, kw_only=True)
+    _word_to_logger: Dict[str, GraphLogger] = attrib(init=False, default=Factory(dict))
+    _rendered_word_hypothesis_pair_ids: Set[str] = attrib(
+        init=False, default=Factory(set)
+    )
+
+    @staticmethod
+    def from_parameters(
+        params: Parameters, *, graph_logger: Optional[GraphLogger] = None
+    ) -> "ProposeButVerifyLanguageLearner":  # type: ignore
+        log_word_hypotheses_dir = params.optional_creatable_directory(
+            "log_word_hypotheses_dir"
+        )
+        if log_word_hypotheses_dir:
+            logging.info("Hypotheses will be logged to %s", log_word_hypotheses_dir)
+
+        rng = Random()
+        rng.seed(params.optional_integer("random_seed", default=0))
+
+        return ProposeButVerifyLanguageLearner(
+            graph_match_confirmation_threshold=params.floating_point(
+                "graph_match_confirmation_threshold"
+            ),
+            graph_logger=graph_logger,
+            log_word_hypotheses_to=log_word_hypotheses_dir,
+            rng=rng,
+            ontology=GAILA_PHASE_1_ONTOLOGY,
+        )
+
+    # observe method
+    def observe(
+        self, learning_example: LearningExample[PerceptionT, LinguisticDescription]
+    ) -> None:
+
+        logging.info("Observation %s", self._observation_num)
+        self._observation_num += 1
+
+        perception = learning_example.perception
+        if len(perception.frames) != 1:
+            raise RuntimeError("Pursuit learner can only handle single frames for now")
```

Comment:
The error message should be "Propose but verify... " instead

Reply:
Changed